Mathlib/Control/Monad/Free.lean

```diff
@@ -0,0 +1,254 @@
+
+@[simp]
+lemma bind_pure : ∀ x : FreeM F α, x.bind (.pure) = x
+  | .pure a => rfl
+  | liftBind op k => by simp [FreeM.bind, bind_pure]
+
+@[simp]
+lemma bind_pure_comp (f : α → β) : ∀ x : FreeM F α, x.bind (.pure ∘ f) = map f x
+  | .pure a => rfl
+  | liftBind op k => by simp only [FreeM.bind, map, bind_pure_comp]
+
+/-- Collapse a `.bind` that follows a `liftBind` into a single `liftBind` -/
+@[simp]
+lemma liftBind_bind (op : F ι) (cont : ι → FreeM F α) (f : α → FreeM F β) :
+    (liftBind op cont).bind f = liftBind op fun x => (cont x).bind f := rfl
+
+instance : LawfulFunctor (FreeM F) where
+  map_const := rfl
+  id_map := id_map
+  comp_map _ _ := comp_map _ _
+
+instance : Monad (FreeM F) where
+
+instance : LawfulMonad (FreeM F) := LawfulMonad.mk'
+  (bind_pure_comp := bind_pure_comp)
+  (id_map := id_map)
+  (pure_bind := pure_bind)
+  (bind_assoc := FreeM.bind_assoc)
+
+/--
+Interpret a `FreeM f` computation into any monad `m` by providing an interpretation
+function for the effect signature `f`.
+
+This function defines the *canonical interpreter* from the free monad `FreeM f` into the target
+monad `m`. It is the unique monad morphism that extends the effect handler
+`interp : ∀ {β}, F β → M β` via the universal property of `FreeM`.
+-/
+protected def liftM {M : Type u → Type w} [Monad M] {α : Type u}
+    (interp : {β : Type u} → F β → M β) : FreeM F α → M α
+  | .pure a => pure a
+  | .liftBind op cont => interp op >>= fun result => (cont result).liftM interp
+
+@[simp]
+lemma liftM_pure {M : Type u → Type w} [Monad M] {α : Type u}
+    (interp : {β : Type u} → F β → M β) (a : α) :
+    (.pure a : FreeM F α).liftM interp = pure a := rfl
+
+@[simp]
+lemma liftM_liftBind {M : Type u → Type w} [Monad M] {α β : Type u}
+    (interp : {γ : Type u} → F γ → M γ) (op : F β) (cont : β → FreeM F α) :
+    (liftBind op cont).liftM interp = (do let b ← interp op; (cont b).liftM interp) := by
+  rfl
+
+@[simp]
+lemma liftM_bind {M : Type u → Type w} [Monad M] [LawfulMonad M] {α β : Type u}
+    (interp : {β : Type u} → F β → M β) (x : FreeM F α) (f : α → FreeM F β) :
+    (x.bind f : FreeM F β).liftM interp = (do let a ← x.liftM interp; (f a).liftM interp) := by
+  induction x generalizing f with
+  | pure a => simp only [pure_bind, liftM_pure, LawfulMonad.pure_bind]
+  | liftBind op cont ih =>
+    rw [FreeM.bind, liftM_liftBind, liftM_liftBind, bind_assoc]
+    simp_rw [ih]
+
+/--
+A predicate stating that `g : FreeM F α → M α` is an interpreter for the effect
+handler `f : ∀ {α}, F α → M α`.
+
+This means that `g` is a monad morphism from the free monad `FreeM F` to the
+monad `M`, and that it extends the interpretation of individual operations
+given by `f`.
+
+Formally, `g` satisfies the two equations:
+- `g (pure a) = pure a`
+- `g (liftBind op k) = f op >>= fun x => g (k x)`
+-/
+structure ExtendsHandler {M : Type u → Type w} [Monad M] {α : Type u}
```

Comment:
What do you think of calling this `IsInterpreterBy f g` or `Interprets g f` or `InterpretsEffect g f`?

Reply:
@tannerduve, @quangvdao